pkg/river/token/builder/builder_test.go

```diff
@@ -0,0 +1,110 @@
+package builder_test
+
+import (
+	"bytes"
+	"testing"
+
+	"github.com/grafana/agent/pkg/river/parser"
+	"github.com/grafana/agent/pkg/river/printer"
+	"github.com/grafana/agent/pkg/river/token"
+	"github.com/grafana/agent/pkg/river/token/builder"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBuilder_File(t *testing.T) {
+	f := builder.NewFile()
+
+	f.Body().SetAttributeTokens("attr_1", []builder.Token{{Type: token.NUMBER, Lit: "15"}})
+	f.Body().SetAttributeTokens("attr_2", []builder.Token{{Type: token.BOOL, Lit: "true"}})
+
+	b1 := builder.NewBlock([]string{"test", "block"}, "")
+	b1.Body().SetAttributeTokens("inner_attr", []builder.Token{{Type: token.STRING, Lit: `"block 1"`}})
+	f.Body().AppendBlock(b1)
+
+	b2 := builder.NewBlock([]string{"test", "block"}, "labeled")
+	b2.Body().SetAttributeTokens("inner_attr", []builder.Token{{Type: token.STRING, Lit: `"block 2"`}})
+	f.Body().AppendBlock(b2)
+
+	expect := format(t, `
+		attr_1 = 15
+		attr_2 = true
+
+		test.block {
+			inner_attr = "block 1"
+		}
+
+		test.block "labeled" {
+			inner_attr = "block 2"
+		}
+	`)
+
+	require.Equal(t, expect, string(f.Bytes()))
+}
+
+func TestBuilder_GoEncode(t *testing.T) {
+	f := builder.NewFile()
+
+	f.Body().SetAttributeValue("null_value", nil)
+	f.Body().AppendTokens([]builder.Token{{token.ILLEGAL, "\n"}})
+
+	f.Body().SetAttributeValue("num", 15)
+	f.Body().SetAttributeValue("string", "Hello, world!")
+	f.Body().SetAttributeValue("bool", true)
+	f.Body().SetAttributeValue("list", []int{0, 1, 2})
+	f.Body().SetAttributeValue("func", func(int, int) int { return 0 })
+	f.Body().SetAttributeValue("capsule", make(chan int))
+	f.Body().AppendTokens([]builder.Token{{token.ILLEGAL, "\n"}})
+
+	f.Body().SetAttributeValue("map", map[string]interface{}{"foo": "bar"})
+	f.Body().SetAttributeValue("map_2", map[string]interface{}{"non ident": "bar"})
+	f.Body().AppendTokens([]builder.Token{{token.ILLEGAL, "\n"}})
+
+	f.Body().SetAttributeValue("mixed_list", []interface{}{
+		0,
+		true,
+		map[string]interface{}{
+			"key_1": true,
+			"key_2": true,
+			"key_3": true,
+		},
+		"Hello!",
+	})
+
+	expect := format(t, `
+		null_value = null
+	
+		num     = 15 
+		string  = "Hello, world!"
+		bool    = true
+		list    = [0, 1, 2]
+		func    = function
+		capsule = capsule("chan int")
```

Comment:
This is kind of a weird way to represent function and capsule values, and needs some thought in the future for how we should display them to users. 

Note that however they get displayed _must_ be valid River syntax, since we need to be able to parse the tokens to format them.